openmls/TEST-DESIGN.md

```diff
@@ -0,0 +1,176 @@
+# OpenMLS testing proposition
+
+The purpose of this document is to define a testing infrastructure and guidelines how to write tests.
+
+## Current status
+
+Here are some factual remarks about the current state of tests.
+
+1. erratic: no clear pattern emerges. It looks like tests have been piled up at different times by different persons
+   without shared, well-established rules. There are very few unit tests, some integration tests in `src/` and some
+   other integration tests in `tests/`
+2. too verbose: some domain are of course complex and hard to test, but some simpler ones require thousands of lines to
+   get tested
+3. not idiomatic: one of the recommended way to write tests in Rust is to write unit tests in the same file as the
+   sources in order to have access to private members. Integration tests (in `tests/`) are more high level and
+   especially
+   used for verifying the public API
+4. Not enough factorization: it is not uncommon to have a file redefining its own set of helpers. A lot of those helpers
+   are duplicated all across the codebase
+5. some tests actually test more than one thing. It makes them harder to write/understand. Also harder to debug when
+   they
+   fail (in which iteration am I ?). This results mostly from the test setup being verbose which incentivizes reuse
+
+## Goals
+
+This proposition aims at:
+
+1. Write less test code
+2. Share a common pattern for writing tests
+3. Have finer grained tests and be quicker to troubleshoot the cause of a failing test
+4. Be idiomatic to welcome easier external contributions
+
+## Rules
+
+To achieve these goals, here are the "rules"/recommendations to set up:
+
+1. Any test expected to fail by altering a valid input SHOULD also keep this valid input and verify that, contrary to
+   the tampered one, it succeeds.
+2. Prefer `.unwrap()` over `.expect("...")`. The latter adds overhead and most of the time does not add enough debugging
+   information. If, on the opposite, an `.expect(...)` can help a developer troubleshooting a failing test with a very
+   complex setup which this `expect` message could help understanding then go for it ! Another
+   argument [here](https://twitter.com/timClicks/status/1584676737572487169).
```

Comment:
As discussed offline: `.unwrap()` is fine in tests, but should be commented on when used for non-trivial checks.